pkg/agent/clean/binding.go

```diff
@@ -0,0 +1,295 @@
+This is technically safe as rancher will recreate any CRB or RB that is deleted that
+should not have been.
+*/
+
+package clean
+
+import (
+	"context"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+	apiv3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/controllers/management/auth"
+	"github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io"
+	v3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	pkgrbac "github.com/rancher/rancher/pkg/rbac"
+	"github.com/rancher/wrangler/pkg/generated/controllers/rbac"
+	v1 "github.com/rancher/wrangler/pkg/generated/controllers/rbac/v1"
+	"github.com/rancher/wrangler/pkg/ratelimit"
+	"github.com/rancher/wrangler/pkg/start"
+	"github.com/sirupsen/logrus"
+	k8srbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	crtbType = "crtb"
+	prtbType = "prtb"
+)
+
+type bindingsCleanup struct {
+	crtbs               v3.ClusterRoleTemplateBindingClient
+	prtbs               v3.ProjectRoleTemplateBindingClient
+	clusterRoleBindings v1.ClusterRoleBindingClient
+	roleBindings        v1.RoleBindingClient
+}
+
+func Bindings() error {
+	logrus.Info("Starting bindings cleanup")
+	if os.Getenv("DRY_RUN") == "true" {
+		logrus.Info("DRY_RUN is true, no objects will be deleted/modified")
+		dryRun = true
+	}
+
+	config, err := clientcmd.BuildConfigFromFlags("", os.Getenv("KUBECONFIG"))
+	if err != nil {
+		panic(err)
+	}
+
+	// No one wants to be slow
+	config.RateLimiter = ratelimit.None
+
+	rancherManagement, err := management.NewFactoryFromConfig(config)
+	if err != nil {
+		return err
+	}
+
+	k8srbac, err := rbac.NewFactoryFromConfig(config)
+	if err != nil {
+		return err
+	}
+
+	starters := []start.Starter{rancherManagement, k8srbac}
+
+	ctx := context.Background()
+	if err := start.All(ctx, 5, starters...); err != nil {
+		return err
+	}
+
+	bc := bindingsCleanup{
+		crtbs:               rancherManagement.Management().V3().ClusterRoleTemplateBinding(),
+		prtbs:               rancherManagement.Management().V3().ProjectRoleTemplateBinding(),
+		clusterRoleBindings: k8srbac.Rbac().V1().ClusterRoleBinding(),
+		roleBindings:        k8srbac.Rbac().V1().RoleBinding(),
+	}
+
+	return bc.clean()
+}
+
+func (bc *bindingsCleanup) clean() error {
+	crtbs, err := bc.crtbs.List("", metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	prtbs, err := bc.prtbs.List("", metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	// The label's key and value changes depending on the rancher version
+	var rancher25 bool
+
+	// Check if we have the updated label, this indicates we are running on rancher 2.5+
+	if len(crtbs.Items) > 0 {
+		if _, ok := crtbs.Items[0].Labels[auth.RtbCrbRbLabelsUpdated]; ok {
+			rancher25 = true
+		}
+	} else if len(prtbs.Items) > 0 {
```

Comment:
should we check both crtbs and prtbs for label existance? "else if" will skip prtbs and decide only on crtbs label

Reply:
If one has it and the other doesn't we are in a really bad state for RBAC, so I just need to check one.